Tests/ConfidenceTests/EventSenderEngineTest.swift

```diff
@@ -67,100 +78,118 @@ final class EventSenderEngineTest: XCTestCase {
+            writeQueue: writeQueue
         )
 
         eventSenderEngine.emit(eventName: "Hello", message: [:], context: [:])
         // TODO: We need to wait for writeReqChannel to complete to make this test meaningful
-        XCTAssertNil(uploader.calledRequest)
+        XCTAssertNil(uploaderMock.calledRequest)
     }
 
     func testRemoveEventsFromStorageOnBadRequest() throws {
         MockedClientURLProtocol.mockedOperation = .badRequest
-        let client = RemoteConfidenceClient(
+        let badRequestUploader = RemoteConfidenceClient(
             options: ConfidenceClientOptions(credentials: ConfidenceClientCredentials.clientSecret(secret: "")),
             session: MockedClientURLProtocol.mockedSession(),
             metadata: ConfidenceMetadata(name: "", version: ""))
 
-        let flushPolicies = [ImmidiateFlushPolicy()]
-        let storage = EventStorageMock()
         let eventSenderEngine = EventSenderEngineImpl(
             clientSecret: "CLIENT_SECRET",
-            uploader: client,
-            storage: storage,
-            flushPolicies: flushPolicies
+            uploader: badRequestUploader,
+            storage: storageMock,
+            flushPolicies: [ImmidiateFlushPolicy()],
+            writeQueue: writeQueue
         )
         eventSenderEngine.emit(eventName: "testEvent", message: ConfidenceStruct(), context: ConfidenceStruct())
         let expectation = expectation(description: "events batched")
-        storage.eventsRemoved{
+        storageMock.eventsRemoved{
             expectation.fulfill()
         }
         wait(for: [expectation], timeout: 2)
 
-        XCTAssertEqual(storage.isEmpty(), true)
+        XCTAssertEqual(storageMock.isEmpty(), true)
     }
 
     func testKeepEventsInStorageForRetry() throws {
+        let expectation = self.expectation(description: "Writes handled")
         MockedClientURLProtocol.mockedOperation = .needRetryLater
-        let client = RemoteConfidenceClient(
+        let retryLaterUploader = RemoteConfidenceClient(
             options: ConfidenceClientOptions(credentials: ConfidenceClientCredentials.clientSecret(secret: "")),
             session: MockedClientURLProtocol.mockedSession(),
             metadata: ConfidenceMetadata(name: "", version: ""))
 
-        let flushPolicies = [ImmidiateFlushPolicy()]
-        let storage = EventStorageMock()
         let eventSenderEngine = EventSenderEngineImpl(
             clientSecret: "CLIENT_SECRET",
-            uploader: client,
-            storage: storage,
-            flushPolicies: flushPolicies
+            uploader: retryLaterUploader,
+            storage: storageMock,
+            flushPolicies: [ImmidiateFlushPolicy()],
+            writeQueue: writeQueue
         )
 
         eventSenderEngine.emit(eventName: "testEvent", message: ConfidenceStruct(), context: ConfidenceStruct())
 
-        XCTAssertEqual(storage.isEmpty(), false)
+        writeQueue.async {
```

Comment:
instead, can we just do the XCTAssert inside `writeQueue.sync`?

Reply:
fixed this